Add support for fstab entries that do not have optional fields

asilveir opened the pull request:
- the last two fstab fields are optional and may be missing.
- the default values were added based on the fstab man page.
- update unit tests.
- add warning for comments at end of valid lines.

This PR aims to fix issue #263

pirat89 commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -87,14 +87,23 @@ def _get_fstab_info(fstab_path):
+                if len(entries) == 5:
+                    entries.append('0')
+
+                if len(entries) == 6:
```
The logic should be different. If you end up with anything else then 6 entries, you should:
- log error & raise stopactorexecutionerror
- or log error and  inhibit upgrade with a msg that fstab seems to be wrong and user should check it. Would be even nice to print the line (or anything that would point what line is problematic at all) so msg will not be confusing so much and user could fix it (or we will be able to investigate it better)

examon replied:
I agree with @pirat89 . It would be nice to log error and inhibit upgrade in case there is a line in fstab with anything else then 6 entries.

pirat89 commented:
@asilveir pls add reference to the issue #263 in the commit message and PR description so it will be closed automatically with merge.

asilveir commented:
I noticed that we are not treating comments at the end of the line I hope I have fixed this too, please @examon and @pirat89 review this too.

I already added a test case for this scenario.

pirat89 commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,32 @@ def _get_fstab_info(fstab_path):
 
+                if '#' in entry:
+                    # Ignore comments at end of lines.
+                    entry = entry.split('#')[0]
```
This seems wrong, as from the documentation:
> Lines starting with '#' are comments.

Regarding the tests, fstab can even contain "#" comments in the end of line (even when this is not documented and it is not recommended to do). So in case this occurs, I would log warning or something. Keep in mind, that current parsing we are doing is not 100% right, as even the hashtag carachter can be used as part of value (e.g. name or password) and do not have to mean it is comment. So I would be careful here and rather than split, I would log error and require remove of the comment from the line or at least log warning.

pirat89 commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,32 @@ def _get_fstab_info(fstab_path):
+                        "at least 4 values and at most 6 this line has {}".format(line, len(entries))
+                    reporting.report_generic(
+                        title='Invalid Fstab', summary=summary, severity='high', flags=['inhibitor']
+                    )
```
You should use remediation report. The summary you wrote is actually more like remediation and it is not telling what is wrong. Summary should be about the error that occur (received X fields, expected...) and remediation instruction what user should do. So, in example I would point users to `man 5 fstab` and explain we expect format documented in the man page with comments on separate lines (ideally).

jmikovic commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,41 @@ def _get_fstab_info(fstab_path):
+                if '#' in entry:
+                    message = "Your /etc/fstab has '#' comment in line {} please verify" \
+                        "if it's correct.".format(line)
+                    api.current_logger().warning(message)
```
You should also strip the comment from the line else you can get inhibitor for valid fstab when it has comment at the end or not get it when there is comment somewhere in the middle but some options are missing

asilveir replied:
I did it before but @pirat89  has commented about it, the '#' is a valid character when you have a password in options for example.

Now we are warning the user when there is the '#'

vinzenz commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,41 @@ def _get_fstab_info(fstab_path):
                     continue
 
+                if '#' in entry:
+                    message = "Your /etc/fstab has '#' comment in line {} please verify" \
```
Please do not use \ for string continuation rather use parenthesis. Logging format strings should usually prefer to use %s.
Strings are supposed to use single quotes not double quotes. 
```suggestion
                    api.current_logger().warning(
                        'Detected in /etc/fstab a "#" comment in line %s please verify if it is correct.', line)
```

vinzenz commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,41 @@ def _get_fstab_info(fstab_path):
+                entries = entry.split()
+
+                if len(entries) not in [4, 5, 6]:
+                    summary = "The fstab configuration file is invalid, you need to fix this to avoid" \
```
Again double quotes and \ for line continuation

vinzenz commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,41 @@ def _get_fstab_info(fstab_path):
+                if len(entries) not in [4, 5, 6]:
+                    summary = "The fstab configuration file is invalid, you need to fix this to avoid" \
+                        "problems in the upgrade process."
+                    remediation = "The file /etc/fstab has a problem on the line: {}, it must have" \
```
Same here

pirat89 commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,46 @@ def _get_fstab_info(fstab_path):
 
+                if '#' in entry:
+                    api.current_logger().warning(
+                        'Detected in /etc/fstab a "#" comment in line %s please verify if it is correct.',
```
```suggestion
                        'Detected a "#" inline comment in the /etc/fstab file which could lead into unexpected behaviour: "%s"',
```

pirat89 commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,46 @@ def _get_fstab_info(fstab_path):
+
                 entry = entry.strip()
                 if not entry:
                     continue
```
That's something that can be done safelly in the beginning of the loop. Whitespaces are not something that would bother us even in case of comments.

fernflower commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,46 @@ def _get_fstab_info(fstab_path):
-                    fs_passno=fs_passno)
+                entries = entry.split()
+
+                if len(entries) not in [4, 5, 6]:
```
`if len(entries) in range(4, 7)` is more self-descriptive imho

fernflower commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,46 @@ def _get_fstab_info(fstab_path):
+                    continue
+
+                data = dict(zip(keys, entries))
+                yield FstabEntry(**data)
```
I have 2 concerns here - kwargs are against our guidelines unless absolutely necessary and data isn't validated anywhere (on the other side - it has never been)

fernflower commented on repos/system_upgrade/el7toel8/actors/storagescanner/libraries/library.py:
```diff
@@ -79,22 +80,46 @@ def _get_fstab_info(fstab_path):
                 if entry.startswith('#'):
                     continue
 
+                if '#' in entry:
```
This will generate a lot of false positives. Either remove this warning and check completely, or check that '#' is the first character of a word.
